text/0000-symbol-name-mangling-v2.md

```diff
@@ -508,170 +795,174 @@ Here are some examples:
-               /         \                             |
-          <path-prefix>  <identifier "quux">         <type> := #1
-             /                                         |
-          <type>                                 <substitution #1>
-            |
-      <absolute-path>
-            |
-      <path-prefix> #1
-           /       \
-<path-prefix> #0   <identifier "Bar">
-      |
-<identifier "foo">
+```rust
+fn demangle_at(&mut pos, mangled, output_string) {
+  if is_backref(*pos, mangled) {
+    // Read the byte offset of the referenced node and
+    // advance `pos` past the backref.
+    let mut referenced_pos = decode(pos, mangled);
+    demangle_at(&mut referenced_pos, mangled, output_string)
+  } else {
+    // do regular demangling
+  }
+}
 ```
 
+Using byte offsets as backref keys (as this RFC does) instead of post-order
+traversal indices (as Itanium mangling does) has the advantage that the
+demangler does not need to duplicate the mangler's substitution indexing logic,
+something that can become quite complex (as demonstrated by the compression
+scheme proposed in the initial version of this RFC).
+
 ### A Note On Implementing Efficient Demanglers
 
-The mangling syntax is constructed in a way that allows for implementing efficient demanglers:
+The mangling syntax is constructed in a way that allows for implementing
+efficient demanglers:
 
- - Mangled names contain information in the same order as unmangled names are expected to contain it. Therefore, a demangler can directly generate its output while parsing the mangled form. There is no need to explicitly instantiate the AST in memory.
+ - Mangled names contain information in the same order as unmangled
+   names are expected to contain it. Therefore, a demangler can directly
+   generate its output while parsing the mangled form. There is no need
+   to explicitly instantiate the AST in memory.
 
- - The same is true for decompression. The demangler can keep a simple array that maps substitution indices to ranges in the already generated output. When it encounters a `<substitution>` in need of expansion, it can just look up corresponding range and do a simple `memcpy`.
+ - The same is true for decompression. Decompression can be done without
+   allocating memory outside of the stack. Alternatively the demangler
+   can keep a simple array that maps back-ref indices to ranges in the
+   already generated output. When it encounters a `<backref>` in need
+   of expansion, it can just look up corresponding range and do a
+   simple `memcpy`.
 
-Parsing, decompression, and demangling can thus be done in a single pass over the mangled name without the need to do dynamic allocation except for the dictionary array.
+Parsing, decompression, and demangling can thus be done in a single pass
+over the mangled name without the need for complex data structures, which
+is useful when having to implement `#[no_std]` or C demanglers.
```

Comment:
This seems to be missing a note about punycode decoding.
(ignore details below if you've read my other descriptions of it)

It requires performing arbitrary-position insertions within a string buffer, driven by instructions decoded from a compressed "instruction stream", with no indication of the final size (no matter which encoding is picked).
Also, those instructions use "unicode scalar value" (i.e. Rust `char`) indexing, so something like `Vec<char>` needs to be used, and later shrunk to a `String`.

One can implement bounded punycode decoding on the stack (i.e. with `[char; MAX_CHARS]`), which gives up and prints the punycode encoding (or does something else) if the number of characters is larger than the constant bound (`128` in my demangler implementation), but there are other workarounds possible (such as specifying a "chunked punycode" system, where the `char`s are split into groups of a fixed size, and are each compressed separately).